README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# contact

<!-- badges: start -->

<!-- badges: end -->

This package is designed to process spatiotemporal data into contact and
social networks, and facilitate network analysis by randomizing
individuals’ movement paths and/or related categorical variables. To use
this package, users need only have a dataset containing spatial data
(i.e., latitude/longitude, or planar x & y coordinates), individual IDs
relating spatial data to specific individuals, and date/time information
relating spatial locations to temporal locations. The functionality of
this package ranges from data “cleaning” via multiple filtration
functions, to spatial and temporal data interpolation, and network
creation and summarization. Functions within this package are not
limited to describing interpersonal contacts. Package functions can also
identify and quantify “contacts” between individuals and fixed areas
(e.g., home ranges, water bodies, buildings, etc.). As such, this
package is an incredibly useful resource for facilitating
epidemiological, ecological, ethological and sociological research.

This package was created and is maintained by members of the [Lanzas
Lab](http://www.lanzaslab.org/) at the North Carolina State University
College of Veterinary Medicine.

## Installation

You can install the released version of contact from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("contact")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("lanzaslab/contact")
```

## Example

*This example is an excerpt from the package vignette.*

Let’s say you want to create a time-aggregated point-location-based
environmental contact networ describing instances when animals,
represented by data points, were in “contact” with a fixed
location/area. As an example of this, here we use *contact* functions to
describe contacts between tracked calves in a single feedlot pen and
their water trough.

``` r
library(contact)

data("calves") #load the calves data set
```

**A.) Ensure all required columns exist in the calves data set (i.e., xy
coordinates, unique individual IDs, dateTime)**

``` r

head(calves)#The calves data set does not have a singular dateTime column. Rather, it has "date" and "time" columns. We must append a dateTime column to the data frame.
#>   calftag     x     y     time       date
#> 1     101 63.38 47.43 00:00:14 05-02-2016
#> 2     101 63.48 46.59 00:00:22 05-02-2016
#> 3     101 62.90 47.07 00:00:26 05-02-2016
#> 4     101 63.27 47.26 00:00:34 05-02-2016
#> 5     101 63.04 46.62 00:00:38 05-02-2016
#> 6     101 63.53 47.24 00:00:42 05-02-2016

calves.dateTime<- contact::datetime.append(x = calves, date = calves$date, time= calves$time, dateTime = NULL, dateFormat = "mdy", dateFake = FALSE, startYear = NULL, tz.in = "UTC", tz.out = NULL, month = FALSE, day = FALSE, year = FALSE, hour = FALSE, minute = FALSE, second = FALSE, daySecond = FALSE, totalSecond = FALSE)
```

**B.) Calculate distances between the water-trough polygon and calves at
each time step.**

First, we must define the location of the water trough. To do this, we
read in point-location data for the water-trough vertices.

``` r

water<- data.frame(x = c(61.43315, 61.89377, 62.37518, 61.82622), y = c(62.44815, 62.73341, 61.93864, 61.67411)) #This is a data frame containing the x and y coordinates of the four trough vertices.
```

As noted in the dist2Area\_df help documention, polygon-vertex
coordinates must be arranged in a particular way. Here we arrange them
accordingly.

``` r

water_poly<-data.frame(matrix(ncol = 8, nrow = 1)) #(ncol = number of vertices)*2
colnum = 0
for(h in 1:nrow(water)){
  water_poly[1,colnum + h] <- water$x[h] #pull the x location for each vertex
  water_poly[1, (colnum + 1 + h)] <- water$y[h] #pull the y location for each vertex
  colnum <- colnum + 1
}
```

Calculate distances between calves and the water polygon at every
timestep.

``` r

water_distance<-contact::dist2Area_df(x = calves.dateTime, y = water_poly, x.id = "calftag", y.id = "water", dateTime = "dateTime", point.x = calves.dateTime$x, point.y = calves.dateTime$y, poly.xy = NULL, parallel = FALSE, dataType = "Point", lonlat = FALSE, numVertices = NULL) #note that the poly.xy and numVertices arguments refer to vertices of polygons in x, not y. Because dataType is "Point," not "Polygon," these arguments are irrelevant here.

head(water_distance)
#>              dateTime totalIndividuals individualsAtTimestep  id
#> 1 2016-05-02 00:00:14               70                    12 101
#> 2 2016-05-02 00:00:14               70                    12 104
#> 3 2016-05-02 00:00:14               70                    12 108
#> 4 2016-05-02 00:00:14               70                    12 111
#> 5 2016-05-02 00:00:14               70                    12 116
#> 6 2016-05-02 00:00:14               70                    12 123
#>   dist.to.water
#> 1      14.32860
#> 2      12.04416
#> 3      32.84068
#> 4      35.58052
#> 5       0.00000
#> 6      47.51557
```

**C.) Identify what SpTh value will allow us to capture 99% of contacts,
defined as instances when point-locations were within 0.333 m of the
water trough, given the RTLS accuracy.**

``` r

SpThValues<-contact::findDistThresh(n1 = 1000, n2 = 1000, acc.Dist1 = 0.5, acc.Dist2 = NULL, pWithin1 = 90, pWithin2 = NULL, spTh = 0.5) #spTh represents the initially-defined spatial threshold for contact

SpThValues #it looks like an adjusted SpTh value of approximately 0.74 m will likely capture 99% of contacts, defined as instances when point-locations were within 0.333 m of the water trough, given the RTLS accuracy. #Note that because these confidence intervals are obtained from distributions generated from random samples, every time this function is run, results will be slightly different. 
#>      mean     5%-CI    10%-CI    15%-CI    20%-CI    25%-CI    30%-CI 
#> 0.7068588 0.7075445 0.7082329 0.7089268 0.7096292 0.7103432 0.7110723 
#>    35%-CI    40%-CI    45%-CI    50%-CI    55%-CI    60%-CI    65%-CI 
#> 0.7118207 0.7125932 0.7133954 0.7142344 0.7151193 0.7160620 0.7170786 
#>    70%-CI    75%-CI    80%-CI    85%-CI    90%-CI    95%-CI    99%-CI 
#> 0.7181923 0.7194380 0.7208727 0.7226002 0.7248454 0.7282912 0.7350257 
#>       max 
#> 1.9878744

CI_99<-unname(SpThValues[21]) #we will use this SpTh value moving forward.
```

**D.) Identify time points when calves were within the re-adjusted SpTh
distance from water trough.**

``` r

water_contacts <- contact::contactDur.area(water_distance, dist.threshold=CI_99,sec.threshold=1, blocking = FALSE, equidistant.time = FALSE, parallel = FALSE, reportParameters = TRUE) #Note that because we are not interested in making a time-aggregated network with > 1 temporal levels, we set blocking = FALSE to reduce processing time.

head(water_contacts)
#>   indiv.id area.id contact.id contactDuration    contactStartTime
#> 1      101   water  101-water               1 2016-05-02 00:47:35
#> 2      101   water  101-water               1 2016-05-02 00:47:43
#> 3      101   water  101-water               1 2016-05-02 00:47:47
#> 4      101   water  101-water               1 2016-05-02 00:47:59
#> 5      101   water  101-water               1 2016-05-02 00:48:15
#> 6      101   water  101-water               1 2016-05-02 00:48:36
#>        contactEndTime distThreshold secThreshold equidistant.time
#> 1 2016-05-02 00:47:35     0.7350257            1            FALSE
#> 2 2016-05-02 00:47:43     0.7350257            1            FALSE
#> 3 2016-05-02 00:47:47     0.7350257            1            FALSE
#> 4 2016-05-02 00:47:59     0.7350257            1            FALSE
#> 5 2016-05-02 00:48:15     0.7350257            1            FALSE
#> 6 2016-05-02 00:48:36     0.7350257            1            FALSE
```


NEWS/NEWS.md

## <!-- NEWS.md is generated from NEWS.Rmd. Please edit that file -->

# contact 1.0.0

  - submitted to CRAN on 10/18/2019

## contact 1.0.1

  - Addressed issues highlighted from CRAN reviewer.
  - Fixed a substantial bug in the “chisq” sub-function of contactTest.

## contact 1.2.0

  - Corrected many bugs in package functions.
  - Improved efficiency (e.g., run time, local memory usage, etc.) of
    package functions
  - Ensured compatability with R 4.0.0.
  - Updated package citation (because our paper was accepted for
    publication\!)
  - Broke contact::contactTest into multiple functions - contactTest is
    now defunct.

## contact 1.2.1

  - Fixed a bug introduced in version 1.2.0, as required by CRAN

## contact 1.2.2

  - Fixed bug associated with temporal sorting of data spanning multiple
    years.
  - Improved efficiency temporal blocking functions.

## contact 1.2.3

  - Fixed bug that forced times in tempAggregate output to begin at
    00:00:00 on the first day in the data set.
  - Added socialEdges function.

## contact 1.2.4

  - Fixed various bugs
  - Realized that the method the findDistThresh function was based on
    was faulty due to a mistake in the paper (Farthing et al. 2020). We
    updated the function to correctly do as intended and updated to
    function documentation to acknowledge the mistake.
  - Drastically improved efficiency of the ntwrkEdges function.

\`\`\`


R/dist2All.R
#' Calculate Distances Between All Individuals
#'
#' Calculates the distance between all tracked individuals at a given timestep. Users can choose whether to calculate distances based on a single point, or polygons representative of individuals' locations. If individuals set dataType == "Polygon", the distance matrix reported describes the shortest distances between polygons' edges (Note that the rgeos::gDistance function is used to obtain these distances).
#'
#' If dataType == "Point," users have the option of setting lonlat == TRUE (by default lonlat == FALSE). lonlat is a logical argument that tells the function to calculate the distance between points on the WGS ellipsoid (if lonlat == TRUE), or on a plane (lonlat == FALSE) (see raster::pointDistance). If lonlat == TRUE, coordinates should be in degrees. Otherwise, coordinates should represent planar ('Euclidean') space (e.g. units of meters).This function is not currently able to calculate distances between polygons on the WGS ellipsoid (i.e., if dataType == "Polygon," lonlat must = FALSE). We aim to address this issue in future versions.
#'
#' Note that if inputting a separate matrix/dataframe with polygon xy coordinates (poly.xy), coordinates must be arranged in the format of those in referencePointToPolygon outputs (i.e., col1 = point1.x, col2 = point1.y, col3 =point2.x, col4 = point2.y, etc., with points listed in a clockwise (or counter-clockwise) order).
#' @param x Data frame or list of data frames containing real-time-location data.  
#' @param id Vector of length nrow(data.frame(x)) or singular character data, detailing the relevant colname in x, that denotes what unique ids for tracked individuals will be used. If argument == NULL, the function assumes a column with the colname "id" exists in x. Defaults to NULL.
#' @param point.x Vector of length nrow(data.frame(x)) or singular character data, detailing the relevant colname in x, that denotes what planar-x or longitude coordinate information will be used. If argument == NULL, the function assumes a column with the colname "x" exists in x. Defaults to NULL.
#' @param point.y Vector of length nrow(data.frame(x)) or singular character data, detailing the relevant colname in x, that denotes what planar-y or lattitude coordinate information will be used. If argument == NULL, the function assumes a column with the colname "y" exists in x. Defaults to NULL.
#' @param dateTime Vector of length nrow(data.frame(x)) or singular character data, detailing the relevant colname in x, that denotes what dateTime information will be used. If argument == NULL, the function assumes a column with the colname "dateTime" exists in x. Defaults to NULL.
#' @param poly.xy Columns within x denoting polygon xy-coordinates. Polygon coordinates must be arranged in the format of those in referencePointToPolygon output. Defaults to NULL.
#' @param elev Vector of length nrow(data.frame(x)) or singular character data, detailing the relevant colname in x, that denotes vertical positioning of each individual in 3D space (e.g., elevation).  If argument != NULL, relative vertical positioning will be incorporated into distance calculations. Defaults to NULL.
#' @param parallel Logical. If TRUE, sub-functions within the dist2All wrapper will be parallelized. Note that this can significantly speed up processing of relatively small data sets, but may cause R to crash due to lack of available memory when attempting to process large datasets. Defaults to TRUE.
#' @param dataType Character string refering to the type of real-time-location data presented in x, taking values of "Point" or "Polygon." If argument == "Point," individuals' locations are drawn from point.x and point.y. If argument == "Polygon," individuals' locations are drawn from poly.xy. Defaults to "Point."
#' @param lonlat Logical. If TRUE, point.x and point.y contain geographic coordinates (i.e., longitude and lattitude). If FALSE, point.x and point.y contain planar coordinates. Defaults to FALSE.
#' @param numVertices Numerical. If dataType == "Polygon," users must specify the number of vertices contained in each polygon. Defaults to 4. Note: all polygons must contain the same number of vertices.
#' @keywords data-processing polygon point location planar GRC
#' @export
#' @examples
#' #load the calves data set
#' data(calves)
#' 
#' #pre-process the data
#' calves.dateTime<-datetime.append(calves, date = calves$date, time = calves$time) #create a dataframe with dateTime identifiers for location fixes.
#' calves.agg<-tempAggregate(calves.dateTime, id = calves.dateTime$calftag, dateTime = calves.dateTime$dateTime, point.x = calves.dateTime$x, point.y = calves.dateTime$y, secondAgg = 10, extrapolate.left = FALSE, extrapolate.right = FALSE, resolutionLevel = "Full", parallel = TRUE, na.rm = FALSE, smooth.type = 1) #smooth locations to 10-second fix intervals. Note that na.rm was set to "FALSE" because randomizing this data set according to Spiegel et al.'s method (see below) requires equidistant time points.
#'
#' #generate empirical time-ordered network edges.
#' calves.dist<-dist2All(x = calves.agg, parallel = TRUE, dataType = "Point", lonlat = FALSE) #calculate distance between all individuals at each timepoint.
#' 
#' More examples will be added later

dist2All<-function(x = NULL, id = NULL, dateTime = NULL, point.x = NULL, point.y = NULL, poly.xy = NULL, elev = NULL, parallel = TRUE, dataType = "Point", lonlat = FALSE, numVertices = 4){

  create.distFrame<- function(x,distMat, indivSeq, idSeq, timestepIndivSeq,time){

    dist = data.frame(matrix(ncol = (length(indivSeq) + 4), nrow = 1))
    colnames(dist) = c("dateTime","totalIndividuals","individualsAtTimestep","id", paste("dist.to.indiv_", idSeq, sep = ""))
    dist$dateTime = time
    dist$totalIndividuals = length(indivSeq)
    dist$individualsAtTimestep = length(timestepIndivSeq)
    dist$id = unlist(unname(x[1]))
    vecID = 1
    column<-as.numeric(unlist(unname(x[2]))) #The column identifier must be numeric. Without the as.numeric call here, apply functions will return the error "Error in distMat[vecID, column] : no 'dimnames' attribute for array" later on.
    col.fill = NULL

    for(j in 1:length(indivSeq)){

      if(isTRUE(indivSeq[j]%in%timestepIndivSeq) == TRUE){

        if(unlist(unname(x[3])) != indivSeq[j]){ #This if statement makes sure that individuals are not reported as being a certain distance from themselves. If the id = indivSeq[j], dist = NA
          col.fill = c(col.fill,distMat[vecID,column]) #column must be must be numeric

        }else{
          col.fill = c(col.fill,NA)
        }
        vecID = (vecID + 1)

      }else{
        col.fill = c(col.fill,NA)
      }
    }

    dist[1,5:ncol(dist)] = col.fill
    return(dist)
  }
  list.breaker1<-function(x,y,id, dateTime, point.x, point.y, poly.xy, elev, parallel, dataType, lonlat, numVertices){ #calculates dist.all or distToArea for randomized input
    input<- data.frame(y[unname(unlist(x[1]))])
    dist.all<- dist.generator1(input,id, dateTime, point.x, point.y, poly.xy, elev, parallel, dataType, lonlat, numVertices)
    return(dist.all)
  }
  dist.generator1<-function(x,id, dateTime, point.x, point.y, poly.xy, elev, parallel, dataType, lonlat, numVertices){
    idVec1=NULL #added in the case that idVec1 isn't created when x isn't specified
    if(length(x) == 0){ #This if statement allows users to input either a series of vectors (id, dateTime, point.x and point.y), a dataframe with columns named the same, or a combination of dataframe and vectors. No matter the input format, a table called "originTab" will be created.
      if(dataType == "point" || dataType == "Point" || dataType == "POINT"){
        originTab = data.frame(id = id, x = point.x, y = point.y, dateTime = dateTime)
      }
      if(dataType == "polygon" || dataType == "Polygon" || dataType == "POLYGON"){

        originTab = data.frame(matrix(ncol = 0, nrow = length(id)))
        originTab$id = id
        colnames(poly.xy)[seq(1,(ncol(poly.xy) - 1),2)] = paste("point",seq(1,(ncol(poly.xy)/2),1),".x", sep = "")
        colnames(poly.xy)[seq(2,ncol(poly.xy),2)] = paste("point",seq(1,(ncol(poly.xy)/2),1),".y", sep = "")
        dateFrame = data.frame(dateTime = dateTime)
        bindlist = list(originTab,poly.xy,dateFrame)
        originTab = data.frame(do.call("cbind", bindlist))
      }
    }

    if(length(x) > 0){ #for some reason using an "else" statement would always result in an originTab table with 0 records...

      if(length(id) > 0){
        if(length(id) == 1 && is.na(match(id[1], names(x))) == FALSE){ #added 1/14 to accompany the list-processing functionality. If x is a list, rather than id being a vector of length(nrow(x)), it may be necessary to designate the colname for intended "id" values (i.e., if the ids in different list entries are different)
          x$id <- x[,match(id, names(x))]
        }else{ #if length(id) > 1
          x$id = id
        }

      }

      idVec1 <- x$id

      if(dataType == "point" || dataType == "Point" || dataType == "POINT"){

        if(length(point.x) > 0){
          if(length(point.x) == 1 && is.na(match(point.x[1], names(x))) == FALSE){ #added 1/14 to accompany the list-processing functionality. If x is a list, rather than point.x being a vector of length(nrow(x)), it may be necessary to designate the colname for intended "point.x" values (i.e., if the x-coordinate values in different list entries are different)
            x$x <- x[,match(point.x, names(x))]
          }else{ #if length(point.x) > 1
            x$x = point.x
          }
        }
        if(length(point.y) > 0){
          if(length(point.y) == 1 && is.na(match(point.y[1], names(x))) == FALSE){ #added 1/14 to accompany the list-processing functionality. If x is a list, rather than point.x being a vector of length(nrow(x)), it may be necessary to designate the colname for intended "point.x" values (i.e., if the x-coordinate values in different list entries are different)
            x$y <- x[,match(point.y, names(x))]
          }else{ #if length(point.y) > 1
            x$y = point.y
          }
        }
        xyFrame1<- data.frame(x = x$x, y = x$y)
      }

      if(dataType == "polygon" || dataType == "Polygon" || dataType == "POLYGON"){

        if(length(poly.xy) > 0){
          if(length(poly.xy) == (numVertices*2) & length(which(is.na(match(poly.xy, names(x)))) == TRUE) == 0){ #added 1/14 to accompany the list-processing functionality. If x is a list, rather than poly.xy being a matrix/dataframe of length(nrow(x)), it may be necessary to designate the colnames for intended coordinate values (i.e., if the xy-coordinate values in different list entries are different)
            xyFrame1<-x[,match(poly.xy,names(x))]
          }else{
            #x[,2:(1 + length(poly.xy))] = poly.xy
            xyFrame1<- data.frame(poly.xy)
          }
        }else{ #if length(poly.xy == 0)
          xyFrame1 <-  x[,(match("point1.x", names(x))):((2*numVertices) + (match("point1.x", names(x)) -1))] #if there is no poly.xy input, the code assumes the input is output from referencePointToPolygon function, and therefore, the first point of interest would be "point1.x"
        }
        colnames(xyFrame1)[seq(1,(numVertices*2),2)] = paste("point",seq(1,numVertices,1),".x", sep = "")
        colnames(xyFrame1)[seq(2,((numVertices*2) + 1),2)] = paste("point",seq(1,numVertices,1),".y", sep = "")
      }

      if(length(dateTime) > 0){

        if(length(dateTime) == 1 && is.na(match(dateTime[1], names(x))) == FALSE){ #added 1/14 to accompany the list-processing functionality. If x is a list, rather than point.x being a vector of length(nrow(x)), it may be necessary to designate the colname for intended "point.x" values (i.e., if the x-coordinate values in different list entries are different)
          x$dateTime <- x[,match(dateTime, names(x))]
        }else{ #if length(dateTime) > 1
          x$dateTime = dateTime
        }

      }
      dateTimeVec<-x$dateTime
      bindlist1<-list(idVec1, xyFrame1, dateTimeVec)
      originTab <- do.call("cbind", bindlist1)
      names(originTab)[c(1,ncol(originTab))]<-c("id", "dateTime")
    }

    #02/01/2019 - Here I'll remove the need for integer ids by creating a separate integ.ID column in originTab
    if(length(idVec1)== 0) { #added to use idVec1 if not created above 
      idVec1 <- originTab$id
      } 
    idVec2=unique(originTab$id) #Added in the case x is not supplied
    originTab$integ.ID<-NA

    for(a in 1:length(idVec2)){
      originTab$integ.ID[which(idVec1 == idVec2[a])] <-as.integer(a)
    }

    originTab$dateTime = as.character(originTab$dateTime)
    
    elev.calc <-FALSE #logical; tells us later if we need to include elevation calculations.
    if(length(elev) > 0){
      elev.calc <-TRUE 
      if(length(elev) == 1 && is.na(match(elev[1], names(x))) == FALSE){ #added 1/14 to accompany the list-processing functionality. If x is a list, rather than elev being a vector of length(nrow(x)), it may be necessary to designate the colname for intended "elev" values (i.e., if the elevs in different list entries are different)
        originTab$elev <- x[,match(elev, names(x))]
      }else{ #if length(elev) > 1
        originTab$elev = elev
      }
    }

    if(dataType == "point" || dataType == "Point" || dataType == "POINT"){

      originTab<-originTab[order(originTab$id, originTab$dateTime),]
      indivSeq = unique(originTab$integ.ID)
      idSeq = unique(originTab$id)
      dateTimeFrame = data.frame(dateTime = unique(originTab$dateTime))
      dist.measurement = lonlat

      dist.process.point <- function(x, originTab, indivSeq, idSeq, dist.measurement, elev.calc){

        time = unlist(unname(x[1]))
        timestep = originTab[which(originTab$dateTime == time),]
        timestepIndivSeq.integ = unique(timestep$integ.ID)
        distMat = raster::pointDistance(timestep[,c(match("x",names(timestep)),match("y",names(timestep)))], timestep[,c(match("x",names(timestep)),match("y",names(timestep)))], lonlat = dist.measurement, allpairs = TRUE)

        if(is.matrix(distMat) == TRUE){ #If there's only one record in timestep, distMat will be a vector, not a matrix, and will only denote the distance between an individual and itself (which isn't useful)
          if(elev.calc == TRUE){
            elevDifference = function(x, elevVec){ #This function creates a matrix showing the difference in elevation between all observed points in the data set (timestep) 
              e1 <- unname(unlist(x[1]))
              de <- as.integer(abs(e1 - elevVec)) 
              return(de)
            }
            elevVec <- unname(unlist(timestep$elev))
            elevFrame <- data.frame(elevVec)
            elev.dif<-apply(elevFrame, 1, elevDifference, elevVec)
            distMat<-distMat + elev.dif #adds the difference in elevations to the linear distances previously calculated. Note that the elev distance units must be in meters, as that is the unit of distMat values. 
          }
          timestepIndivSeqFrame = data.frame(id = unique(timestep$id), colnum = seq(1,length(unique(timestep$id)),1), timestepIndivSeq.integ) #Note that colnum will not necessarily be the same as timestepIndivSeq.integ because originTab has been previously ordered
          distTab = data.frame(data.table::rbindlist(apply(timestepIndivSeqFrame,1,create.distFrame,distMat,indivSeq, idSeq, timestepIndivSeq.integ,time)))
          return(distTab)
        }
      }


      if (parallel == TRUE){
        cl<-parallel::makeCluster(parallel::detectCores())
        distTab<-parallel::parApply(cl, dateTimeFrame, 1, dist.process.point, originTab, indivSeq,idSeq, dist.measurement, elev.calc)
        parallel::stopCluster(cl)
      }else{
        distTab = apply(dateTimeFrame, 1, dist.process.point, originTab, indivSeq,idSeq, dist.measurement, elev.calc)
      }
      dist.all = data.frame(data.table::rbindlist(distTab))
    }

    if(dataType == "polygon" || dataType == "Polygon" || dataType == "POLYGON"){

      originTab<-originTab[order(originTab$id, originTab$dateTime),]
      naVec<-which(is.na(originTab[,match("point2.x", names(originTab))]) == TRUE) #the referencePointtoPolygon function will create some observations that are not complete polygons (i.e., only the point1 coordinates are recorded). This identifies those observations, so that they may be removed. If they are not removed, they will cause errors later.
      if(length(naVec) > 0){
        originTab <- originTab[-naVec,]
      }
      indivSeq = unique(originTab$integ.ID)
      idSeq <- unique(originTab$id)
      dateTimeFrame = data.frame(dateTime = unique(originTab$dateTime))

      dist.process.poly <- function(x, originTab, indivSeq, idSeq, numVertices, elev.calc){
        create.poly1<-function(x,y, numVertices){
          polygon = unlist(c(y[unlist(unname(x[1])),c(2:(1 + (numVertices*2)),2:3)])) #note that when making a Spatial Polygon you need to close the polygon by repeating the first point at the end (Note that this also means you need to add a "+ 1" to nrow = numVertices below)
          spatPoly = sp::Polygons(list(sp::Polygon(matrix(polygon, nrow = (numVertices + 1), ncol = 2, byrow = TRUE))),y$integ.ID[unlist(unname(x[1]))])
          return(spatPoly)
        }

        time = unlist(unname(x[1]))
        timestep = originTab[which(originTab$dateTime == time),]

        if(nrow(timestep) > 1){ #If there's only one record in timestep, any distMat produced will only denote the distance between an individual and itself (which isn't useful), so it's more time efficient to skip forward nrow(timestep <= 1)

          timestepIndivSeq.integ = unique(timestep$integ.ID)
          makePolyFrame<-data.frame(seq(1,length(timestepIndivSeq.integ),1))
          spatialPolygons <- apply(makePolyFrame,1,create.poly1,timestep,numVertices)
          sPolys = sp::SpatialPolygons(spatialPolygons,timestepIndivSeq.integ) #note that the second part of this argument must be an integer. Otherwise, it will return the following error: Error: is.integer(pO) is not TRUE
          distMat = rgeos::gDistance(sPolys, byid = TRUE)
          
          if(elev.calc == TRUE){
            elevDifference = function(x, elevVec){ #This function creates a matrix showing the difference in elevation between all observed points in the data set (timestep) 
              e1 <- unname(unlist(x[1]))
              de <- as.integer(abs(e1 - elevVec)) 
              return(de)
            }
            elevVec <- unname(unlist(timestep$elev))
            elevFrame <- data.frame(elevVec)
            elev.dif<-apply(elevFrame, 1, elevDifference, elevVec)
            distMat<-distMat + elev.dif #adds the difference in elevations to the linear distances previously calculated. Note that the elev distance units must be in meters, as that is the unit of distMat values. 
          }
          
          timestepIndivSeqFrame = data.frame(id = unique(timestep$id), colnum = seq(1,length(unique(timestep$id)),1), timestepIndivSeq.integ) #Note that colnum will not necessarily be the same as timestepIndivSeq.integ because originTab has been previously ordered
          distTab = data.frame(data.table::rbindlist(apply(timestepIndivSeqFrame,1,create.distFrame,distMat,indivSeq, idSeq, timestepIndivSeq.integ,time)))
          return(distTab)
        }
      }

      if (parallel == TRUE){
        cl<-parallel::makeCluster(parallel::detectCores())
        distTab<-parallel::parApply(cl, dateTimeFrame, 1, dist.process.poly, originTab, indivSeq, idSeq, numVertices, elev.calc)
        parallel::stopCluster(cl)
      }else{
        distTab = apply(dateTimeFrame, 1, dist.process.poly, originTab, indivSeq, idSeq, numVertices, elev.calc)
      }
      dist.all = data.frame(data.table::rbindlist(distTab))
    }
    return(dist.all)
  }

  if(is.data.frame(x) == FALSE & is.list(x) == TRUE){ #1/15 added the "is.data.frame(x) == FALSE" argument because R apparently treats dataframes as lists.
    breakFrame<- data.frame(seq(1,length(x),1))
    list.dist <- apply(breakFrame, 1, list.breaker1,y = x,id, dateTime, point.x, point.y, poly.xy, elev, parallel, dataType, lonlat, numVertices) #in the vast majority of cases, parallelizing the subfunctions will result in faster processing than parallelizing the list processing here. As such, since parallelizing this list processing could cause numerous problems due to parallelized subfunctions, this is an apply rather than a parApply or lapply.
    return(list.dist)

  }else{ #if(is.list(x) == FALSE)

    frame.dist<- dist.generator1(x,id, dateTime, point.x, point.y, poly.xy, elev, parallel, dataType, lonlat, numVertices)

    return(frame.dist)
  }
}
